### ISSUE TITLE
Nomad 1.7.0 does not honor env VAULT_TOKEN

### ISSUE THREAD

**Issue by Artanicus** [hashicorp/nomad#19342]
### Nomad version
```
Nomad v1.7.0
BuildDate 2023-12-07T08:28:54Z
Revision e4150e9703f3be6ee2339f0e45ff0801186e022b
```
### Operating system and Environment details
N/A

### Issue
Vault 1.7.0 is not capable of reading the vault token from the env variable VAULT_TOKEN. This is a regression from 1.6.4 and not mentioned as a breaking change in the release, thus I'm assuming it's a bug.

### Reproduction steps
Simplified repro:
`docker run --rm -it -e NOMAD_SKIP_DOCKER_IMAGE_WARN="trustme" -e VAULT_TOKEN="legit" hashicorp/nomad:1.7.0 agent -server -data-dir=/tmp -vault-enabled`
vs.
`docker run --rm -it -e NOMAD_SKIP_DOCKER_IMAGE_WARN="trustme" -e VAULT_TOKEN="legit" hashicorp/nomad:1.6.4 agent -server -data-dir=/tmp -vault-enabled`

#### Expected Result
Server starts

#### Actual Result
On 1.7.0:
> ==> Error starting agent: server setup failed: Failed to setup Vault client: Vault token must be set
    2023-12-07T14:26:33.833Z [INFO]  nomad: shutting down server
    2023-12-07T14:26:33.833Z [ERROR] nomad: failed to setup Vault client: error="Vault token must be set"
    2023-12-07T14:26:33.833Z [ERROR] agent: error starting agent: error="server setup failed: Failed to setup Vault client: Vault token must be set"

**Comment by tgross** [comment]
Hi @Artanicus! Sorry about that. I can reproduce this. 

We rebuilt the Vault config parsing as part of the new multi-cluster support for Nomad Enterprise. When I grabbed the `VAULT_TOKEN` environment variable here [`command.go#L283`](https://github.com/hashicorp/nomad/blob/v1.7.0/command/agent/command.go#L283) it's getting applied to a version of the config data that's after we've tried to "merge" any configurations and that produces a copy. So the environment variable is being applied to the original and not the copy that actually gets used by the agent.

Should be an easy fix. We're likely to have a fast-follow patch for 1.7.1. In the meantime, you can use the `vault.token` config or the command line argument `-vault-token` as a workaround.

**Comment by tgross** [comment]
We've shipped a patch for this bug in Nomad 1.7.1.
